### ISSUE TITLE
Add the new kernel scheduler load balance algorithm to ipvs-scheduler option on kube-proxy ipvs-mode

### ISSUE THREAD

**Issue by mcarbonneaux** [kubernetes/kubernetes#114619]
The ipvs mode `ipvs-scheduler` limit the scheduler algorithm to :
```
              rr - Round Robin: distributes jobs equally amongst the available real servers.

              wrr - Weighted Round Robin: assigns jobs to real servers  proportionally  to  there
              real  servers'  weight.  Servers with higher weights receive new jobs first and get
              more jobs than servers with lower weights. Servers with equal weights get an  equal
              distribution of new jobs.

              lc - Least-Connection: assigns more jobs to real servers with fewer active jobs.

              wlc  -  Weighted Least-Connection: assigns more jobs to servers with fewer jobs and
              relative to the real servers' weight (Ci/Wi). This is the default.

              lblc - Locality-Based Least-Connection: assigns  jobs  destined  for  the  same  IP
              address to the same server if the server is not overloaded and available; otherwise
              assign jobs to servers with fewer jobs, and keep it for future assignment.

              lblcr - Locality-Based Least-Connection with Replication: assigns jobs destined for
              the  same  IP  address  to  the  least-connection node in the server set for the IP
              address. If all the node in the server set are over loaded, it picks up a node with
              fewer  jobs  in  the  cluster  and  adds it in the sever set for the target. If the
              server set has not been modified for the specified time, the most  loaded  node  is
              removed from the server set, in order to avoid high degree of replication.

              dh  -  Destination Hashing: assigns jobs to servers through looking up a statically
              assigned hash table by their destination IP addresses.

              sh - Source Hashing: assigns jobs  to  servers  through  looking  up  a  statically
              assigned  hash  table  by their source IP addresses.  This scheduler has two flags:
              sh-fallback, which enables fallback to a different server if  the  selected  server
              was  unavailable,  and  sh-port,  which  adds  the  source  port number to the hash
              computation.

              sed - Shortest Expected Delay: assigns an incoming  job  to  the  server  with  the
              shortest  expected  delay. The expected delay that the job will experience is (Ci +
              1) / Ui if  sent to the ith server, in which Ci is the number of jobs  on  the  the
              ith server and Ui is the fixed service rate (weight) of the ith server.

              nq - Never Queue: assigns an incoming job to an idle server if there is, instead of
              waiting for a fast one; if all  the  servers  are  busy,  it  adopts  the  Shortest
              Expected Delay policy to assign the job.
```

https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/apis/config/types.go#L226

now in new kernel have this new algorithm  : 
```
              fo  -  Weighted  Failover:  assigns  an incoming job to the server with the highest
              weight that is currently available.

              ovf - Weighted Overflow: assigns an incoming job to the  server  with  the  highest
              weight  that  is  currently  available  and  overflows  to  the  next  when  active
              connections exceed the node's  weight.  Note  that  this  scheduler  might  not  be
              suitable for UDP because it only uses active connections.

              mh  -  Maglev  Hashing:  assigns  incoming  jobs  based  on Google's Maglev hashing
              algorithm, providing an almost equal share of jobs to each real server and provides
              minimal  disruption. When the set of real servers changes, a connection will likely
              be sent to the same real server as it was before.  This scheduler  has  two  flags:
              mh-fallback,  which  enables  fallback to a different server if the selected server
              was unavailable, and mh-port, which  adds  the  source  port  number  to  the  hash
              computation.
```

the `mh` are specialy interesting... that be added in kernel 4.18...

**Comment by thockin** [comment]
This seems non-controversial on its face, but a) how do we know whether the kernel supports it or not; and b) if every node makes independent decisions, doesn't this basically become random?

/assign uablrek

**Comment by uablrek** [comment]
a) It should be the user's responsibility to ensure that the configured ipvs sheduler is available on all nodes

This sounds rude, but please read https://github.com/kubernetes/kubernetes/issues/108579#issuecomment-1125675626 for some rationale (and https://github.com/kubernetes/kubernetes/issues/108579#issuecomment-1159770477).

Exception _may_ be the `rr` which is the default, but an installation with `ipvs` but _without_ the `rr` sheduler is _extremely_ unlikely.

If that relaxing of the sheduler pre-test is acceptable, I would say that _any_ current and future ipvs sheduler shall be allowed. Just treat them as a string.

But it should be documented in the release-notes and some place else, but I don't know where.

**Comment by uablrek** [comment]
b) For all shedulers that make decisions based on local state, yes. (that includes `rr` btw)

Well, not really random ... For instance if `lc` is used and a new endpoint is addred, then all ipvs instances will get an "empty" target and will select it for new connections and may overload the newcomer, or it will become "statistically correct" over time. 

But IMHO users should be allowed to try. Some sheduler may be very useful in their environment. That will (already has) of course spawn some issues when it doesnt work as expected (for the reason in b) for instance), but allowing all shedulers will not change that.

**Comment by uablrek** [comment]
So, my proposal is:
- Drop the scheduler module check
- Treat the configured scheduler as a string in the ipvs proxier
- Document

**Comment by uablrek** [comment]
The module check can be replaced with a _function_ check, e.g. insert/delete an ipvs dummy rule.

**Comment by aojea** [comment]
> The module check can be replaced with a _function_ check, e.g. insert/delete an ipvs dummy rule.

people is trying to reduce privileges on containers these days, does it requires more privileges than today or is the same?

**Comment by uablrek** [comment]
The same. Kube-proxy configures ipvs already. This would only add an extra probing insert/delete on start. Even on first start (if ipvs is empty) with some care.

**Comment by uablrek** [comment]
Could even be less privileges is "modprobe" requires any privileges (now or in the future)

**Comment by uablrek** [comment]
### Off-topic question about Maglev in ipvs

> the mh are specialy interesting... that be added in kernel 4.18...

@mcarbonneaux What is your use-case?

I have wanted to check the `mh` sceduler in ipvs, but havn't got time yet. The [Maglev](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/44824.pdf) load-balancer is stateless, but `ipvs` is not (well, at least not normally).

A problem Maglev tries to solve is if the number of Application Servers (AS) (endpoints in K8s) changes, the updated hash buckets must be as few as possible. But if the load-balancer is stateful and stores the connections (like ipvs or the Linux conntrack used in proxy-mode=iptables), no existing connections are affected, no matter what hash algorithm is used for new connections.

So, for both proxy-mode ipvs and iptables existing connection are not affected by adding or removing endpoints.

But... if external traffic is routed to different nodes, e.g. a GW rehashes ECMP targets after a cluster scaling, _then_ traffic is affected.

This is easily tested:
* Setup a service of type LoadBalancer
* Route external traffic to the loadBalancerIP to node A
* Setup a number of TCP connections
* Re-route external traffic to the loadBalancerIP to node B

All connections are lost (I have tested). But if the load-balancer would have been a stateless Maglev _no connections would be lost._

Google Maglev stores connections but only as a cache and getting the best of stateful, existing connections never lost on AS scaling, and stateless, the load-balancing tier can be scaled with minimum impact. This _can_ be the case in ipvs `mh`, but is it?

**Comment by thockin** [comment]
Lars: proposal SGTM.  IPVS impl is pretty much your call to make :)

**Comment by uablrek** [comment]
@thockin  I have started working on a PR to remove the module checks. You are the only active reviewer and I will try my best to divide it into small reviewable commits as you wrote in [google groups](https://groups.google.com/g/kubernetes-sig-network/c/iCZlKNCmazg/m/Po2ymPExBQAJ) :smile:

**Comment by uablrek** [comment]
PR to remove the module checks https://github.com/kubernetes/kubernetes/pull/114669.

I also intend to remove the constraint on ipvs schedulers, which will fix this issue. If you can create a VS with your configured scheduler you can use it, whatever that scheduler may be.

**Comment by uablrek** [comment]
> Google Maglev stores connections but only as a cache and getting the best of stateful, existing connections never lost on AS scaling, and stateless, the load-balancing tier can be scaled with minimum impact. This can be the case in ipvs mh, but is it?

Looks like it is. Please read https://blog.cloudflare.com/high-availability-load-balancers-with-maglev/

I have tested "mh" in kube-proxy (with a patch) and first problem is that is hashes on address only by default, so a traffic generator using the same source all the time will end up on the same endpoint. This can be configured but will require special treatment in kube-proxy.

**Comment by thockin** [comment]
@uablrek Should we close this now that #114669 is in?

**Comment by uablrek** [comment]
No, https://github.com/kubernetes/kubernetes/pull/114669 just removes the module ckeck. The validator still restricts the usable schedulers

**Comment by mcarbonneaux** [comment]
> What is your use-case?

tu add afinity without using states. is the finality of maglev to avoid perf on managing state... and is more robust on target lost than other consistant hashing method...

> I have tested "mh" in kube-proxy (with a patch) and first problem is that is hashes on address only by default, so a traffic generator using the same source all the time will end up on the same endpoint. This can be configured but will require special treatment in kube-proxy.

is normal is what you whant when you use maglev... to have afinity on ip src... but without need of storing states... is afinity by consistant hashing...

cilium use this algorithm too.... 
https://cilium.io/blog/2020/11/10/cilium-19/#maglev

**Comment by mcarbonneaux** [comment]
the google research paper on the algorythm:
https://static.googleusercontent.com/media/research.google.com/fr//pubs/archive/44824.pdf

**Comment by uablrek** [comment]
@mcarbonneaux If you can build and use an own `kube-proxy` please try PR https://github.com/kubernetes/kubernetes/pull/114878.

I have tested and "mh" can be configured, and with many src-addresses load-balancing works. However if external ingress traffic is re-routed from one K8s node to another, all connections are lost, same as "rr".

**Comment by mcarbonneaux** [comment]
with ipvs the node who receive the traffic is considered a director, if you LB to other node... this node must have default route to the director because of nat masquerading... 

when you use MH/RR with Direct Server Return avoid that problem (no more nat masquerading)... the node respond to the client directly (no need to return back to the first node who receive the traffic).

but you need to tunnel the traffic from node who receive the request to the node you forward the traffic (cloud flare use fou protocol for that)...
https://blog.cloudflare.com/high-availability-load-balancers-with-maglev/

idont know if kubeproxy already support that sort of configuration...

cilium for LB DSR + Maglev use tcp option field to store the Virtual IP that receive the traffic in the packet and nat the destination with the pod ip before sending to the node who host the node  (that work in full routing mode), and the node respond to the client directly in replacing thaire ip with the Virtual IP from tcp option field.... in that way no state needed... no tunneling... no complex network configuration....

some has been done on kube-router project:
- https://github.com/cloudnativelabs/kube-router/issues/167
- https://github.com/cloudnativelabs/kube-router/issues/166 

and i think you must use `externalTrafficPolicy: Local` in conjonction with MH/RR to use Direct Server Return if i understand whell...
- https://dev.to/kcdchennai/direct-server-return-with-kubernetes-454l

**Comment by sondinht** [comment]
@uablrek, Regarding "mh" algorithm, I submitted PR #114745 a few weeks ago which adds the new load distribution method to kube-proxy. As this commit supports 'mh" with hashing on srcIP and srcPort by default, connections from the same srcIP with different srcPorts will be distributed equally to all EndPoints. If users want to have an affinity on srcIP, they can use "mh" with "Session Affinity" option enabled.

The PR is targeted for the main use-cases below:
- Connection failover from a worker node to another.
- Scale-in/out worker nodes 
- Rerouting of GW or External LoadBalancer changes receiving worker nodes of existing traffic flows.

I would need to rebase the commit after your fix for this issue PR #114878.
PR #114745 is now waiting triage state, please @thockin verify if it is reasonable to add this new algorithm to kube-proxy.

**Comment by uablrek** [comment]
> i dont know if kubeproxy already support that sort of configuration...

No, and it probably never will.

The proper (only) way to get this into `kube-proxy` is to write a [Kubernetes Enhancement Proposal](https://github.com/kubernetes/enhancements/blob/master/keps/README.md) (KEP). To [quote myself](https://kubernetes.io/blog/2021/10/18/use-kpng-to-write-specialized-kube-proxiers/):

> It is a tedious process to get new features into the kube-proxy and it is not unlikely that they will be rejected, so to write a specialized proxier may be the only option.

**Comment by aojea** [comment]
I would also like to see a demo or e2e test working with this setup, it doesn't seem obvious from the comments that this will work out of the box ... The project Kubernetes, as Lars correctly mentions , use KEPs to discuss and document feature, I agree with him and I think this will require to write a KEP

**Comment by uablrek** [comment]
> @mcarbonneaux What is your use-case?

What I meant was your use-case _in K8s_. Having worked most of 2021 with the Maglev hash algorithm when developing the [nfqueue-loadbalancer](https://github.com/Nordix/nfqueue-loadbalancer/) I know what it can do, but I also know K8s and I think I can say right now;

**To allow the "mh" in kube-proxy=ipvs will not give you what you hope for.**

**Comment by sondinht** [comment]
@aojea, @uablrek I'm new to this community. Thanks for your suggestion about the process of adding a new feature to kube-proxy. I'll prepare a KEP for the "mh" load distribution.

**Comment by uablrek** [comment]
But in any case I would like PR https://github.com/kubernetes/kubernetes/pull/114878 to be merged. For sanity reasons if not others.

The `FlagSourceHash` for scheduler=mh can/should be added in a separate issue+PR. This will allow users to test "mh" in kube-proxy, because;

> To allow the "mh" in kube-proxy=ipvs will not give you what you hope for.

Nobody will take my word for it :smile:

**Comment by uablrek** [comment]
@sondinht I don't want to sound discouraging but IMO you should not spend too much time on a KEP before presenting it to get some early feedback. I can't see how "mh" can provide the statelessness unless DSR is used, and that may be _very hard_ to get into the `kube-proxy`. So, I suggest you describe a use-case in the KEP, and what it takes to support it, and present it on a bi-weekly sig/network meeting. And don't worry, people are nice there :smile:

**Comment by sondinht** [comment]
Perhaps DSR is not needed to use with "mh" since I cannot really see any problem with nat masquerading of ipvs if subsequence packets are sent to the second node.
  
With net/ipv4/vs/sloppy_* enabled, ipvs on the second node accepts the packets as an existing connection and then keeps forwarding them to the original endpoint thanks to the mh algorithm.  IPVS of course performs snat and "remember" the client IP of the packets as if they are sent to the first node. When receiving replies from the endpoint, ipvs on the second node will be able to forward the replies to the client with the cached client IP in its NAT table.

Please correct me if I'm missing something in this regard.

**Comment by uablrek** [comment]
```
# Traffic goes to k8s-node-A
client (src=client,dst=VIP) -> kube-proxy(src=k8s-node-A,dst=pod) -> pod
# Traffic re-directed to k8s-node-B
client (src=client,dst=VIP) -> kube-proxy(src=k8s-node-B,dst=pod) -> pod does not know this is the same TCP-connection
client (RST) <- kube-proxy(RST) <- pod(RST)
```

So even if the kube-proxy forwards packets to the same POD (assuming the same mh configuration in node A and B) the receiving POD will reject with a TCP-RST packet.

**Comment by uablrek** [comment]
Since you have written PR https://github.com/kubernetes/kubernetes/pull/114745 you should be able to test this with you own-built kube-proxy. For me all connections are teminated when re-directing ingress traffic from one K8s node to another.

**Comment by sondinht** [comment]
Thanks @uablrek for your clarification. Now I could see this problem is on the pod side, not ipvs. It seems that having DSR does not help to solve the problem.

Though, I suppose this issue will not happen with the following scenarios:
- UDP traffic
- Traffic is sent to Cluster-IP instead of NodeIP

**Comment by thockin** [comment]
> Nobody will take my word for it

Me neither :)

> I don't want to sound discouraging but IMO you should not spend too much time on a KEP before presenting it to get some early feedback.

Agree - you will have the best experience if do this iteratively.  Start with an email to the SIG list.  Then a doc discussing the peoblem and high-level solution details.  THEN a KEP makes sense.

**Comment by mcarbonneaux** [comment]
> To add generic support for DSR in K8s is not simple.

cilium assigne vip also but more higher in the kernel stack with ebpf...

ipvs can do that, helped with some config on each node...

**Comment by sondinht** [comment]
> Is this an acceptable way?
> 
> 1. Get [Ipvs: remove the scheduler validation #114878](https://github.com/kubernetes/kubernetes/pull/114878) merged. This will allow "mh" to be configured as a ipvs scheduler, but only with L3 hashing (no ports).

Agree with the approach. #114878 fixes the raised issue that it enables all distribution methods supported by ipvs in kube-proxy (including "mh" one).

> 2. Write a new issue like "Ipvs scheduler=mh only hashes on addresses, not ports", and implement it in a new (small) PR with FlagSourceHash as @sondinht have already done in [Add ipvs support of Maglev load distribution method to kube-proxy #114745](https://github.com/kubernetes/kubernetes/pull/114745)

A separate issue for enabling source port hash when using 'mh' sounds reasonable. This is not only applicable for "mh", but also for the "sh" method which kube-proxy is currently supporting as stated in k8s documents.

> 3. Close this issue as "done" since "mh" can be configured and [Add ipvs support of Maglev load distribution method to kube-proxy #114745](https://github.com/kubernetes/kubernetes/pull/114745) as it's implemented in another way.

Agree.

> 4. Initiate the work towards a KEP to make "mh" work as you want it to work. Including a K8s use-case, DSR (if needed in some form), setting of sysctl's, etc.
> 
> I leave item 4. to you @sondinht and @mcarbonneaux . Item 1-3 will enable "mh" but because of NAT, connections are still lost when traffic is re-directed from one node to another (this can be tested without too much effort. I have done it).

Sure. I'll need to verify the issue further and check if any easy way possible to avoid the SNAT issue of ipvs scheduler located on different nodes. IMHO, replacing VIP (virtual IP) with NodeIP seems to be the first option to overcome the issue, but we need to verify its feasibility and impact in k8s.

**Comment by uablrek** [comment]
>> To allow the "mh" in kube-proxy=ipvs will not give you what you hope for.
>>    Nobody will take my word for it
>
> Me neither :)

:smile: Ok, ok, please let me rephrase:

To just allow "mh" in the _current_ ipvs proxier will not give you what you hope for.

So, https://github.com/kubernetes/kubernetes/pull/114878 is not enough.

**Comment by kylepl** [comment]
So the base maglev support, and some variations for including different elements in the hash are done.

> 4. Initiate the work towards a KEP to make "mh" work as you want it to work. Including a K8s use-case, DSR (if needed in some form), setting of sysctl's, etc.

Was this ever done or pushed through? I couldn't find anything more about it. Basically was:

> So even if the kube-proxy forwards packets to the same POD (assuming the same mh configuration in node A and B) the receiving POD will reject with a TCP-RST packet.

solved? With the ideal outcome being if packets are split across kube-proxy nodes for a given TCP connection, the connection still works.

I'm investigating this a bit, see more discussion at https://github.com/kubernetes/kubernetes/issues/132710.
